### ISSUE TITLE
0.54 obscures required inputs within scatter block

### ISSUE THREAD

**Issue by dpark01** [dnanexus/dxWDL#84]
This is great progress. One bug to note: I'm finding that required input parameters are no longer accessible from compiled workflows in 0.54 if they are buried in a scatter.

Example workflow: project-F8PQ6380xf5bK0Qk0YPjB17P:workflow-F8Zj0B80xf5jg5PgP5fq5jpb

[Original WDL](https://github.com/broadinstitute/viral-ngs/blob/master/pipes/WDL/workflows/demux_plus.wdl)

Although dxWDL compiles, the tasks called within the scatter block don't have their required inputs exposed, either via `dx run --help`, `dx describe`, or `dx describe --json`.

Some example missing inputs are `spikein.spikein_db`, `spades.trim_clip_db`, etc.

0.52 would render these inputs as `stage_2.spikein_spikein_db` and `stage_2.spades_trim_clip_db`

I'll also note that the compile itself [complains about missing inputs in a new way](https://travis-ci.org/broadinstitute/viral-ngs/jobs/313080784), almost sounding like an error message, but the compile is successful, and for the non-scatter tasks, those inputs are properly exposed.

**Comment by orodeh** [comment]
@dpark01 (@mlin)  The issue with the definitions inside a `scatter` (or an `if` block), is that they are compiled into an auxiliary applet. They don't have DNAx natural settings, like stages do. I did expose them in earlier releases, but I was not very happy with the resulting complexity. Can you live without it, or is this an absolute must? 

As food for thought, I may reimplement scatters as a sub-workflow. Those do not take *any* implicit arguments. 

The warning language is too strong; it should be phrased better. 
```
Call is missing a compulsory argument.
Workflow doesn't supply required input reads_unmapped_bam to call kraken 
    call metagenomics.kraken  {
         ^
```
I will fix it.

**Comment by dpark01** [comment]
There are inputs (both required and optional) within the scatter that I will certainly want to be able to configure at launch time (whether CLI or web UI). How best to pass these through I don't know. The previous approach (0.52) seemed fine to me, carrying them up to the auxiliary applet and adding prefixes.

**Comment by orodeh** [comment]
@dpark01 and @mlin, For what it's worth, I have thought about this long and hard. 

The general approach is name mangling; replacing a dot `.` with an underscore `_`, or  a triple underscore `___`. What happens though, if we have more than one level of nesting? A scatter within a scatter, and inside of it if blocks? You can replace the dots with underscores, but are likely to cause name collisions. 

Propagating unspecified arguments to become workflow level inputs runs the risk of an explosion in the number of workflow level inputs. This is difficult to present from a UI perspective, and adds the challenge of linking each workflow input to its stage gear-box.

I would rather not decide on this now, but rather wait for more user feedback. I definitely hear you, that you like and want this feature. As a workaround, is to add variables where you need them.

**Comment by mlin** [comment]
Yes, we will keep discussing how dxWDL should expose these. FWIW a more general discussion also ongoing in OpenWDL: https://github.com/openwdl/wdl/pull/161

@orodeh -- one further point of emphasis after re-reading the thread -- if I understand correctly above, the tasks being called inside the scatter have **required** inputs being left implicit, which dxWDL is not exposing as it reports successful compilation. So one gets a compiled workflow that cannot be run successfully. Did I misunderstand something?

**Comment by orodeh** [comment]
Right, that's a bug. I'll make a note of it, and fix it.

**Comment by mlin** [comment]
Tried this out on v0.56, so far so good.

This is minor but I think you could still restore more of the old language from "Note: workflow doesn't supply required input xxx to stage yyy; leaving corresponding DNAnexus workflow input unbound." The current "Call is missing a compulsory argument" still sounds a bit like a hard error message, while per https://github.com/openwdl/wdl/pull/161 this is a fairly idiomatic pattern in WDL. The original message could probably be made more concise, but it communicates clearly that it's an FYI "Note", and that the compiler is automatically dealing with it.

**Comment by dpark01** [comment]
Yes, this looks resolved to me on 0.56!
